sys/include/net/ieee802154/radio.h

```diff
@@ -0,0 +1,1085 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @defgroup    drivers_ieee802154_hal IEEE802.15.4 Radio Hardware Abstraction Layer
+ * @ingroup     drivers
+ * @brief       This is a Hardware Abstraction Layer for IEEE802.15.4 compatible
+ *              radios.
+ *
+ * @author      José I. Alamos <jose.alamos@haw-hamburg.de>
+ */
+
+#ifndef IEEE802154_RADIO_H
+#define IEEE802154_RADIO_H
+
+#include <stdbool.h>
+#include "iolist.h"
+#include "sys/uio.h"
+#include "byteorder.h"
+#include "net/eui64.h"
+
+/**
+ * @brief Forward declaration of the radio ops structure.
+ */
+typedef struct ieee802154_radio_ops ieee802154_radio_ops_t;
+
+typedef enum {
+    /**
+     * @brief the transceiver received a valid ACK with the frame pending bit
+     */
+    TX_FLAGS_FRAME_PENDING  =(1 << 0),
+    /**
+     * @brief the transceiver ran out of retransmission retries
+     */
+    TX_FLAGS_NO_ACK         =(1 << 1),
+    /**
+     * @brief the CSMA-CA algorithm or Auto CCA failed to measure a clear channel
+     */
+    TX_FLAGS_MEDIUM_BUSY    =(1 << 2),
+} ieee802154_tx_status_t;
+
+/**
+ * @brief IEEE802.15.4 Radio capabilities
+ *
+ * These flags represent the hardware capabilities of a given device.
+ */
+typedef enum {
+    /**
+     * @brief the device support frame retransmissions with CSMA-CA
+     */
+    IEEE802154_CAP_FRAME_RETRIES,
+    /**
+     * @brief the device support ACK timeout interrupt
+     */
+    IEEE802154_CAP_IRQ_ACK_TIMEOUT,
+    /**
+     * @brief the transceiver supports transmission with automatic CCA
+     */
+    IEEE802154_CAP_AUTO_CCA,
+    /**
+     * @brief the device supports Auto ACK
+     */
+    IEEE802154_CAP_AUTO_ACK,
+    /**
+     * @brief the device supports the IEEE802.15.4 2.4 GHz band
+     */
+    IEEE802154_CAP_24_GHZ,
+    /**
+     * @brief the device support the IEEE802.15.4 Sub GHz band
+     */
+    IEEE802154_CAP_SUB_GHZ,
+    /**
+     * @brief the device reports when the transmission is done
+     */
+    IEEE802154_CAP_IRQ_TX_DONE,
+    /**
+     * @brief the device reports the start of a frame (SFD) when received.
+     */
+    IEEE802154_CAP_IRQ_RX_START,
+    /**
+     * @brief the device reports the end of the CCA procedure
+     */
+    IEEE802154_CAP_IRQ_CCA_DONE,
+    /**
+     * @brief the device supports going to RX after TX done.
+     */
+    IEEE802154_CAP_TXRX,
+    /**
+     * @brief the device provides the number of retransmissions
+     */
+    IEEE802154_CAP_FRAME_RETRIES_INFO,
+} ieee802154_rf_caps_t;
```

Comment:
Should this also list different PHY modes (O-QPSK, MR-O-QPSK, MR-O-OFDM, …)?

Reply:
I would say so

Reply:
is it possible to group these modes? (e.g by IEEE 802.15.4 version, SubGHz, etc)

Reply:
Not really.
All PHY modes are specified for all bands, although restrictions apply to channel width, spreading factor, etc based on region.

(btw.: [IEEE 802.15.4-2020](https://ieeexplore.ieee.org/document/9144691) is on [Sci-Hub](https://sci-hub.se/))

Reply:
Ok, I think we can simply add them as caps. It won't hurt